### ISSUE TITLE
Standardize frame and image terminology and their related methods in `ImagingExtractor`

### ISSUE THREAD

**Issue by h-mayorquin** [catalystneuro/roiextractors#121]
##TL;DR 
In videos with multiple channels (three colors) the terms **image** and **frame** can mean an object with shape `(n_channels, rows, columns)` or an object with shape `(row, columns)`. I think we should decided which one is which and enforce their related methods to return objects with that shape by default.

Particularly, I think we should change `get_frames` to return an object of size  `(n_channels, rows, columns)` as now it mostly (as, in most of the methods in the instance classes) returns objects of size `(row, columns)`.

Here, I don't mind about the order, just that the n_channels dimension should be there.
## Context
The base class for imaging extractor has the following abstract methods:

https://github.com/catalystneuro/roiextractors/blob/35accb9c6b25838dfe209cd7ea626b99518d9120/src/roiextractors/imagingextractor.py#L29-L35

Looking at the first method (there is a channel in the argument) it seems that the original intention was for frame to mean an object of size `(row, columns)` and some of the tests that I recently silenced are consistent with this assumption:

https://github.com/catalystneuro/roiextractors/blob/35accb9c6b25838dfe209cd7ea626b99518d9120/src/roiextractors/testing.py#L181-L188

This has not been a problem so far because all of our `io` tests only have data with one channel but the recently added `NumpyMemmapImagingExtractor` method can handle videos with three channels and the test above was supprsed. 

My personal preference is for frame to mean  `(n_channels, rows, columns)`  and that image means `(rows, columns)`  (you can see that that's a [common understanding](https://stackoverflow.com/questions/13033278/image-size-python-opencv) of the word and that `openCV` used to have like that). As mentioned above, following this convention would mean that `get_frames` should return objects of size  `(n_channels, rows, columns)` by default which is a small change in most of the classes. 

That said, maybe there is a better terminology or something that I am missing.

**Comment by CodyCBakerPhD** [comment]
To be clear though, only the capability of the `NumpyMemmapImagingExtractor` involves multiple channels. In practice, these data types are most often grayscale (I haven't yet seen any true Imaging data that is in color), including the current Seidemann project that uses this binary format.

So another option may be just for the moment to raise `NotImplementedErrors` for when there are more more than one color channel.

**Comment by h-mayorquin** [comment]
@CodyCBakerPhD 

I see, so we should expect to see all the future formats to only have a gray channel?

Raise the error where? You mean, assuming that there is only one channel and that all videos are `(n_frames, rows, columns)` (or dim < 4) and raise an error otherwise somewhere?

I would rather just have the convention of the extra dimension for indicating the channels as this allows the code to be future proof. But maybe I am missing some cost of that allowance.

**Comment by CodyCBakerPhD** [comment]
@h-mayorquin More like, all past formats have only used one channel but there's infrastructure on roiextractors evidencing an expectation that future versions might support multiple channels.

Sbx: https://github.com/catalystneuro/roiextractors/blob/master/src/roiextractors/extractors/sbximagingextractor/sbximagingextractor.py#L72-L80 all map to `nChans=1`, which is what is used to set the frame shape below.

Tiff also infers number of channels from the 'video' frames, but we've not seen any actual TIFF data that is anything other than a single channel.

So in the way, TIFF and the other non-sbx `ImagingExtractors` get their total shape from the video array shape after `__init__`, so it could *technically* be any 4-D shape with any range of channels.

> Raise the error where? You mean, assuming that there is only one channel and that all videos are (n_frames, rows, columns) (or dim < 4) and raise an error otherwise somewhere?

Probably in the parent at whatever common `super()` calls there are across interfaces. It's not that we can't support it, it's that we've no data to prototype it on aside from synthetic types, which we also aren't testing either.

> I would rather just have the convention of the extra dimension for indicating the channels as this allows the code to be future proof. But maybe I am missing some cost of that allowance.

I think that was the original thought process. I'm proposing keep it built in, but only offering `NotImplementedError`-free operation until we can actually test to make sure it's working as expected.

**Comment by h-mayorquin** [comment]
I don't understand what are you proposing to keep built in. What are you referring too?

The method `get_frames` and `get_video` right now return data with the following shape:
 `(rows, columns)` and `(num_frames, rows, columns)`
 
 Or a permutation of that. 

I am proposing that they should return something with the shape:

`(n_channels, rows, columns)` and `(n_frames, n_channels, rows, columns)` 

Or a permutation of that.
Where n_channels would be 1 (for all of working examples) and 3 for the present synthetic data and future testing data. 

That way, `frame` means all the channels of an image as in common usage and not a specific channel that is going to be selected randomly (currently 0) when the methods are inherited. Plus, we will not need to re-write our functions when new three dimensional data comes around.

**Comment by CodyCBakerPhD** [comment]
> I don't understand what are you proposing to keep built in.

I'm just saying whenever the value of `n_channels` is not `==1`, we should probably let the user know that we haven't yet tested any functionality and they should consider sharing said data so we can build around it.

**Comment by CodyCBakerPhD** [comment]
> My personal preference is for frame to mean (n_channels, rows, columns) and that image means (rows, columns) (you can see that that's a [common understanding](https://stackoverflow.com/questions/13033278/image-size-python-opencv) of the word and that openCV used to have like that).

Also from that discussion, and from the behavior of OpenCV, when there are multiple channels they generally show up as the outermost dimension. I would prefer then if we standardized as `standard_shape = (frames, rows, columns, channels)` whenever `n_channels > 1` and simply `(frames, rows, columns)` when `channels==1`. This prevents redundancy of having an axis of length one on the boundary by allowing us to drop it when uninformative.

Also, it's interesting to see rows/cols referred to as 'height' and 'width', what do you think about that? rows/cols are references to a particular orientation of an array/matrix, but height/width refers to physical portions. Both are technically true since each image is a discretized visual grid of the physical space, but the former refers to the image as a model of the reality and the latter refers to the reality being modelled.

**Comment by h-mayorquin** [comment]
@CodyCBakerPhD 

> I'm just saying whenever the value of n_channels is not ==1, we should probably let the user know that we haven't yet tested any functionality and they should consider sharing said data so we can build around it.
- If you want to throw a warning or an assertion at the right abstract class level when `num_channels==3` am fine with it. For the ones that are tested like NumpyMemmapset we can use a flag to avoid the warning or assertion.
> Also from that discussion, and from the behavior of OpenCV, when there are multiple channels they generally show up as the outermost dimension. I would prefer then if we standardized as standard_shape = (frames, rows, columns, channels) whenever n_channels > 1 and simply (frames, rows, columns) when channels==1. This prevents redundancy of having an axis of length one on the boundary by allowing us to drop it when uninformative.
- As previously discussed I am fine with standardizing the inner format in which we operate in this repo.  Plus, having the channel dimension at the end is the format that I am used to work with so even better! Where we disagree is on the point of suppressing the channel dimension when `num_channels==1`. I just prefer consistency and I am totally fine with redundancy. Specially for the internal presentation I want to know that the videos and frames will have a consistent shape for my operations. My experience working with images is that not knowing the shapes of your objects appear requires you having checks all over the place and reshaping to make the code consistent for images of different channels. 
> Also, it's interesting to see rows/cols referred to as 'height' and 'width', what do you think about that? rows/cols are references to a particular orientation of an array/matrix, but height/width refers to physical portions. 
- Concerning nomenclature of `rows` and `columns` I think that rows and columns are a more faithful descriptor of the data structures that we are using while doing image processing. That is, in an image processing context I prefer references to the data structures themselves instead of references to the represented objects.  But that's just my preference. I can see good arguments for making it the other way around and as long as it is not `x` and `y` (which is ambiguous) I think is fine.

What I really want though and the point of this issue is that I think that `get_frame` and `get_video` should not return only channel by default though. Which I guess is related to the second point above.

**Comment by CodyCBakerPhD** [comment]
Another thing to add to this discussion is keyword standardization.

`get_frames` uses `frame_idxs`, whereas `frame_to_time` uses `frames`. But their role is the functions are exactly the same. Haven't exhaustively searched the entire library for other similar uses.

I'd say we go with `frames` for both. @h-mayorquin What do you think?

**Comment by h-mayorquin** [comment]
> Another thing to add to this discussion is keyword standardization.
> 
> `get_frames` uses `frame_idxs`, whereas `frame_to_time` uses `frames`. But their role is the functions are exactly the same. Haven't exhaustively searched the entire library for other similar uses.
> 
> I'd say we go with `frames` for both. @h-mayorquin What do you think?

Good catch, @CodyCBakerPhD . I will keep an eye open for inconsistencies like this.

**Comment by h-mayorquin** [comment]
Change `rows` and `columns` to `num_rows` and `num_columns`.

**Comment by CodyCBakerPhD** [comment]
@h-mayorquin We're working on getting the readthedocs updated and this kind of thing would be great to have a section on. Could you add that to your to-do list?

For example, this tutorial for EXTRACT https://github.com/schnitzer-lab/EXTRACT-public/blob/master/Learning%20materials/Lecture%20Tutorials/Tutorial%201%20-%20Starting%20code/example_tutorial.m

indicates the data storage as `height` x `width`, which we as `rows` x `cols` (https://github.com/catalystneuro/roiextractors/blob/master/src/roiextractors/extractors/schnitzerextractor/extractsegmentationextractor.py#L131, though not terribly obvious to read from that code, also not very educational since the example data is square...) but then this will get transposed to NWB standard, which is `width` x `height`... It can get very confusing and it would be great to have a nicely documented showcase for it

**Comment by h-mayorquin** [comment]
Moving to another issue.
